core/src/main/java/org/axonframework/config/MessageMonitorFactoryBuilder.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2010-2017. Axon Framework
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.config;
+
+import org.axonframework.common.Assert;
+import org.axonframework.messaging.Message;
+import org.axonframework.monitoring.MessageMonitor;
+import org.axonframework.monitoring.NoOpMessageMonitor;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+
+/**
+ * Class used by {@link DefaultConfigurer} to maintain the configuration for Message Monitors and create the function
+ * used in the {@link Component}.
+ */
+class MessageMonitorFactoryBuilder {
+
+    // Comparator to ensure the keys are processed in a reasonably stable order in #getFactoryForType
+    // Added comparing on the class' classloader as a class can be loaded multiple times by different classloaders
+    private Comparator<Class<?>> classComparator =
+            Comparator.comparing((Class<?> c) -> c.getName())
+                      .thenComparingInt((Class<?> c) -> c.getClassLoader().hashCode());
+
+    private Map<String, SortedMap<Class<?>, MessageMonitorFactory>> forNameFactories = new HashMap<>();
```

Comment:
Is this a map containing a map per name because two different components can have the same name?

Reply:
Or is it because you could add a message monitor for example for the name 'commandBus', which you tie to the `AsynchronousCommandBus`, `SimpleCommandBus`, `DistributedCommandBus` etc?

Reply:
The former: it is allowed (but not best practice) for two different components to have the same name.